frontend/Rhyolite/Frontend/Modal/Class.hs

```diff
@@ -0,0 +1,44 @@
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeFamilies #-}
+
+module Rhyolite.Frontend.Modal.Class where
+
+import Control.Monad.Reader (MonadReader (ask), ReaderT (..))
+import Control.Monad.Trans (MonadTrans (lift))
+import Reflex (Event, Reflex, EventWriterT)
+import Obelisk.Route.Frontend (RoutedT, askRoute, runRoutedT)
+
+-- | The class of monads supporting a 'tellModal' operation which will open a modal
+--   that stays on top of all other content.
+class HasModal t m where
+  -- If 'm' is the monad that supports 'tellModal' then 'ModalM m' is the monad that the modal itself is in,
+  -- which, notably, probably doesn't support 'tellModal'.
+  type ModalM m :: * -> *
+
+  -- | Opens a modal when the given event fires. The event carries a function which:
+  --     * takes a "close" event triggered when the user signifies that they want to close the modal,
+  --     * builds content in 'ModalM m',
+  --     * returns a "close" event which will be used to actually close the modal.
+  --
+  --   For example, a modal may choose not to be closable by simply ignoring it's input and returning 'never'.
+  tellModal :: Event t (Event t () -> ModalM m (Event t ())) -> m ()
+
+  default tellModal :: (MonadTrans f, m ~ f m', HasModal t m', Monad m', ModalM (f m') ~ ModalM m') => Event t (Event t () -> ModalM m (Event t ())) -> m ()
+  tellModal = lift . tellModal
+
+instance (Monad m, Reflex t, HasModal t m) => HasModal t (EventWriterT t w m) where
+  type ModalM (EventWriterT t w m) = ModalM m
```

Comment:
I understand the special case `type ModalM (ModalT t m) = m`, but not this one

Reply:
This one means that ModalM will not be able to "pass through" `EventWriterT`. E.g. `runEventWriterT $ tellModal $ \_ -> tellEvent x` doesn't work because the monad inside `tellModal` has stripped off the `EventWriterT`.

I believe we can make this use case work though. We've just never needed it.